spark3/src/main/java/org/apache/iceberg/actions/Spark3MigrateAction.java

```diff
@@ -0,0 +1,144 @@
+ * Table in the same location with the same identifier. Once complete the identifier which
+ * previously referred to a non-iceberg table will refer to the newly migrated iceberg
+ * table.
+ */
+class Spark3MigrateAction extends Spark3CreateAction {
+  private static final Logger LOG = LoggerFactory.getLogger(Spark3MigrateAction.class);
+  private static final String BACKUP_SUFFIX = "_BACKUP_";
+
+  Spark3MigrateAction(SparkSession spark, CatalogPlugin sourceCatalog, Identifier sourceTableName) {
+    super(spark, sourceCatalog, sourceTableName, sourceCatalog, sourceTableName);
+  }
+
+  @Override
+  public Long execute() {
+    Table icebergTable;
+
+    // Move source table to a new name, halting all modifications and allowing us to stage
+    // the creation of a new Iceberg table in its place
+    String backupName = sourceTableName().name() + BACKUP_SUFFIX;
+    Identifier backupIdentifier = Identifier.of(sourceTableName().namespace(), backupName);
+    try {
+      destCatalog().renameTable(sourceTableName(), backupIdentifier);
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException("Cannot find table to migrate", e);
+    } catch (TableAlreadyExistsException e) {
+      throw new IllegalArgumentException("Cannot rename migration source to backup name", e);
+    }
+
+    StagedTable stagedTable = null;
+    try {
+      Map<String, String> newTableProperties = new ImmutableMap.Builder<String, String>()
+          .put(TableCatalog.PROP_PROVIDER, "iceberg")
+          .putAll(JavaConverters.mapAsJavaMapConverter(v1SourceTable().properties()).asJava())
+          .putAll(tableLocationProperties(sourceTableLocation()))
+          .putAll(additionalProperties())
+          .build();
+
+      stagedTable = destCatalog().stageCreate(Identifier.of(
+          destTableName().namespace(),
+          destTableName().name()),
+          v1SourceTable().schema(),
+          sourcePartitionSpec(),
+          newTableProperties);
+
+      icebergTable = ((SparkTable) stagedTable).table();
+
+      if (!icebergTable.properties().containsKey(TableProperties.DEFAULT_NAME_MAPPING)) {
+        applyDefaultTableNameMapping(icebergTable);
+      }
+
+      String stagingLocation = icebergTable.location() + "/" + ICEBERG_METADATA_FOLDER;
+
+      LOG.info("Beginning migration of {} using metadata location {}", sourceTableName(), stagingLocation);
+
+      Some<String> backupNamespace = Some.apply(backupIdentifier.namespace()[0]);
+      TableIdentifier v1BackupIdentifier = new TableIdentifier(backupIdentifier.name(), backupNamespace);
+      SparkTableUtil.importSparkTable(spark(), v1BackupIdentifier, icebergTable, stagingLocation);
+
+      stagedTable.commitStagedChanges();
+    } catch (Exception e) {
+      LOG.error("Error when attempting perform migration changes, aborting table creation and restoring backup", e);
+
+      try {
+        if (stagedTable != null) {
+          stagedTable.abortStagedChanges();
```

Comment:
Since the changes are only staged, this should happen after restoring the backup table, in case there is a failure in the abort.

Reply:
Sure, that order does make more sense

Reply:
Is this one actually resolved?

Reply:
Well, it is in try block, never mind.

Reply:
I switched it back after we decided we catch the exception and just log it. But we can move it back